### ISSUE TITLE
Efficient BLI with trained BERT

### ISSUE THREAD

**Issue by oxmwh** [yaoyiran/BLI-Reading-List#2]
Hi 
I'm new to BLI, I thought to ask your opinion.
I have bilingually trained LLMs like BERT or GPT style and I want a word by word translation.
What methods achieve a half descent result while being very efficient?
Since I'm doing this on a huge library, and need it to be fast.
Cheers

**Comment by yaoyiran** [comment]
Hi, generally there are two types of methods for your choice. 

(1) I would recommend embedding-based retrieval [ContrastiveBLI](https://github.com/cambridgeltl/ContrastiveBLI) using BERT-extracted embeddings + static embeddings, optionally with improved performance by additional reranking via [BLICEr](https://github.com/cambridgeltl/BLICEr/graphs/traffic). These methods need model fine-tuning and also applicable for smaller models such as BERT. The inference is very efficient.

(2) I would also recommend [Prompt4BLI](https://github.com/cambridgeltl/prompt4bli) (if you have a seed dictionary) or [SAIL-BLI](https://github.com/cambridgeltl/sail-bli) (if you do not have a seed dictionary). These two methods prompt GPT-like LLMs and do not need any model fine-tuning. They can derive stronger performance compared to embedding-based approaches. But usually they really depend on quite larger and more state-of-the-art LLMs in order to achieve good results, and thus are less efficient during inference.

From your question, I guess (1) might be more suitable since you care inference rather than training efficiency. If your library is in a special domain not seen during LLM pretraining, both (1) and (2) methods need model fine-tuning I think.

**Comment by yaoyiran** [comment]
Usually BLI inference efficiency is a concern only in some academic scenarios. When building real-world applications, the inference efficiency is also nothing to care about, since dictionary inference can be done offline over your whole vocabulary, and BLI inference is not longer needed thereafter. After a single scan over the whole vocabulary, the inferred bilingual dictionary is stored as e.g., a python dict, and word translations can be retrieval in constant time.

**Comment by oxmwh** [comment]
Ah, I see... I've looked at those techniques, I guess I've confused something else with BLI.
although its very similar, maybe you can give me some insights. here is the problem I'm trying to solve:
we have a sentence in Spanish, and we try to get a word by word interlinear translation:
```
la          hermosa            flor             roja
the       beautiful            flower        red
```
but we can't simply use a dictionary since each words translation depends on the context:
```
the cat chased the MOUSE
the MOUSE is connected to USB 3.0
```
and we want to do this on thousands of books.
so maybe using a language model to get context aware embeddings for each word and map that to the target language?

**Comment by yaoyiran** [comment]
I think it is known as "word alignment" task, different from BLI. Their motivations are different. BLI more of addresses extremely low-resource scenarios where no parallel sentences are available (at least simulate low-resource setups with high-resource languages). Word alignment does have enough parallel data either with or without annotation, and mainly hope to get some alignments. BLI can only be used as a baseline for word alignment, which is very efficient. If you find the performance is not really bad, then maybe it's worth proposing some ways to integrate BLI into word alignment approaches.
